packages/eslint-plugin/src/rules/non-nullable-type-assertion-style.ts

```diff
@@ -43,22 +43,42 @@ export default util.createRule({
       return tsutils.unionTypeParts(type);
     };
 
+    const couldBeNullish = (type: ts.Type): boolean => {
+      if (type.flags & ts.TypeFlags.TypeParameter) {
+        const constraint = type.getConstraint();
+        return constraint == null || couldBeNullish(constraint);
+      } else if (tsutils.isUnionType(type)) {
```

Comment:
Generally if you're checking individual flags in a type union, you'll want to use `tsutils.unionTypeParts`, in case the union type is a union of unions.

```ts
type AndAOne = 'a' | 'b';
type AndATwo = 'c' | 'd';
type Joined = AndAOne | AndATwo | 'e';
```

However, `tsutils.isTypeFlagSet` _should_ do the trick for you, I think.